### PR TITLE
Lbank2: add contract API endpoints

### PR DESCRIPTION
**Pull request by Dan-krm**
Added contract API endpoints to lbank2:
closes: #18817
https://www.lbank.com/en-US/docs/contract.html#basic-data

Added available swap support Lbank2:

### fetchMarkets:
```
lbank2.fetchMarkets ()
2023-08-10T01:53:45.206Z iteration 0 passed in 1491 ms

              id |            symbol |        base | quote | settle |      baseId | quoteId | settleId | type |  spot | margin |  swap | future | option | active | contract | linear | inverse | contractSize | expiry | expiryDatetime |  strike | optionType |                         precision |                                                                                              limits
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
        btc_usdt |          BTC/USDT |         BTC |  USDT |        |         btc |    usdt |          | spot |  true |  false | false |  false |  false |   true |          |        |         |              |        |                |         |            |    {"amount":0.0001,"price":0.01} |                                        {"leverage":{},"amount":{"min":0.0001},"price":{},"cost":{}}
        eth_usdt |          ETH/USDT |         ETH |  USDT |        |         eth |    usdt |          | spot |  true |  false | false |  false |  false |   true |          |        |         |              |        |                |         |            |    {"amount":0.0001,"price":0.01} |                                        {"leverage":{},"amount":{"min":0.0001},"price":{},"cost":{}}
        bch_usdt |          BCH/USDT |         BCH |  USDT |        |         bch |    usdt |          | spot |  true |  false | false |  false |  false |   true |          |        |         |              |        |                |         |            |     {"amount":0.0001,"price":0.1} |                                        {"leverage":{},"amount":{"min":0.0001},"price":{},"cost":{}}
...
        LINKUSDT |    LINK/USDT:USDT |        LINK |  USDT |   USDT |        LINK |    USDT |     USDT | swap | false |  false |  true |  false |  false |   true |     true |   true |         |            1 |        |                |         |            |            {"amount":1,"price":1} |  {"leverage":{},"amount":{"min":0.01,"max":2000000},"price":{"min":0.2,"max":0.2},"cost":{"min":5}}
         LTCUSDT |     LTC/USDT:USDT |         LTC |  USDT |   USDT |         LTC |    USDT |     USDT | swap | false |  false |  true |  false |  false |   true |     true |   true |         |            1 |        |                |         |            |            {"amount":1,"price":1} | {"leverage":{},"amount":{"min":0.001,"max":10000},"price":{"min":0.05,"max":0.05},"cost":{"min":5}}
         XLMUSDT |     XLM/USDT:USDT |         XLM |  USDT |   USDT |         XLM |    USDT |     USDT | swap | false |  false |  true |  false |  false |   true |     true |   true |         |            1 |        |                |         |            |          {"amount":0.1,"price":1} |    {"leverage":{},"amount":{"min":1,"max":10000000},"price":{"min":0.2,"max":0.2},"cost":{"min":5}}
1046 objects
```

### fetchTime:
```
lbank2.fetchTime ()
2023-08-11T21:33:47.420Z iteration 0 passed in 300 ms

1691789627950
```

### fetchTickers:
```
lbank2.fetchTickers (BTC/USDT:USDT)
2023-08-15T03:41:54.782Z iteration 0 passed in 274 ms

{
  'BTC/USDT:USDT': {
    symbol: 'BTC/USDT:USDT',
    timestamp: undefined,
    datetime: undefined,
    high: 29446.5,
    low: 29359.2,
    bid: undefined,
    bidVolume: undefined,
    ask: undefined,
    askVolume: undefined,
    vwap: 14856.078285806609,
    open: 29419.5,
    close: 29361.8,
    last: 29361.8,
    previousClose: undefined,
    change: -57.7,
    percentage: -0.1961284182260065,
    average: 29390.65,
    baseVolume: 3972.8364,
    quoteVolume: 59020768.5751021,
    info: {
      prePositionFeeRate: '0.000053',
      volume: '3972.8364',
      symbol: 'BTCUSDT',
      highestPrice: '29446.5',
      lowestPrice: '29359.2',
      openPrice: '29419.5',
      markedPrice: '29361.2',
      turnover: '59020768.5751021',
      lastPrice: '29361.8'
    }
  }
}
```

### fetchOrderBook:
```
lbank2.fetchOrderBook (BTC/USDT:USDT, 3)
2023-08-16T03:45:27.763Z iteration 0 passed in 279 ms

{
  symbol: 'BTC/USDT:USDT',
  bids: [ [ 29234.1, 22.2585 ], [ 29234, 6.6415 ], [ 29233.9, 16.4994 ] ],
  asks: [ [ 29234.2, 12.2615 ], [ 29234.3, 15.1974 ], [ 29234.4, 12.7726 ] ],
  timestamp: 1692157527762,
  datetime: '2023-08-16T03:45:27.762Z',
  nonce: undefined
}
```

### CODE REVIEW

File: ts/src/lbank2.ts [hunk 1]
```diff
@@ -852,14 +872,17 @@ export default class lbank2 extends Exchange {
...
          * @description query for balance and get the amount of funds available for trading or funds locked in orders
+         * @see https://www.lbank.info/en-US/docs/index.html#asset-information
+         * @see https://www.lbank.info/en-US/docs/index.html#account-information
+         * @see https://www.lbank.info/en-US/docs/index.html#get-all-coins-information
          * @param {object} [params] extra parameters specific to the lbank2 api endpoint
          * @returns {object} a [balance structure]{@link https://docs.ccxt.com/en/latest/manual.html?#balance-structure}
          */
         await this.loadMarkets ();
         let method = this.safeString (params, 'method');
         if (method === undefined) {
```
Comment (carlosmiei) [inline ts/src/lbank2.ts:883]: @Dan-krm the `type` (in params or options) should decide the endpoint here

File: ts/src/lbank2.ts [hunk 2]
```diff
@@ -618,7 +636,7 @@ export default class lbank2 extends Exchange {
         params = this.omit (params, 'method');
         if (method === undefined) {
             const options = this.safeValue (this.options, 'fetchTrades', {});
-            method = this.safeString (options, 'method', 'publicGetTrades');
+            method = this.safeString (options, 'method', 'spotPublicGetTrades');
```
Comment (carlosmiei) [inline ts/src/lbank2.ts:639]: @Dan-krm same here, why aren't we checking the type?

Comment (Dan-krm) [reply ts/src/lbank2.ts:639]: @carlosmiei These methods that you're pointing out only have spot support just multiple different endpoints that can be used for spot that I carried over from the previous implementation. I haven't added the swap methods yet, in this PR I just added the swap endpoints and adjusted the naming of the spot endpoints and added some info to the docstrings.

File: ts/src/lbank2.ts [hunk 3]
```diff
@@ -449,15 +633,74 @@ export default class lbank2 extends Exchange {
...
          */
         await this.loadMarkets ();
-        const request = {
-            'symbol': 'all',
-        };
-        const response = await this.publicGetTicker24hr (this.extend (request, params));
+        let market = undefined;
+        if (symbols !== undefined) {
+            symbols = this.marketSymbols (symbols);
+            market = this.market (symbols[0]);
```
Comment (carlosmiei) [inline ts/src/lbank2.ts:646]: what if the user calls fetchtickers with an empty array?

File: ts/src/lbank2.ts [hunk 4]
```diff
@@ -278,37 +292,90 @@ export default class lbank2 extends Exchange {
...
-        //      [
-        //          {
-        //              symbol: 'snx3s_usdt',
-        //              quantityAccuracy: '2',
-        //              minTranQua: '0.01',
-        //              priceAccuracy: '6'
-        //          }
-        //     ]
+        const spotMarkets = await this.fetchSpotMarkets (params);
+        const swapMarkets = await this.fetchSwapMarkets (params);
```
Comment (carlosmiei) [inline ts/src/lbank2.ts:347]: @Dan-krm we should always use Promise.all to resolve both promises "in parallel"

**Comment by carlosmiei** [comment]
LGTM

```
 p lbank2 fetchTicker "BTC/USDT:USDT"
Python v3.10.9
CCXT v4.0.54
lbank2.fetchTicker(BTC/USDT:USDT)
{'ask': None,
 'askVolume': None,
 'average': 26016.9,
 'baseVolume': 29237.4234,
 'bid': None,
 'bidVolume': None,
 'change': -197.0,
 'close': 25918.4,
 'datetime': None,
 'high': 26128.5,
 'info': {'highestPrice': '26128.5',
          'lastPrice': '25918.4',
          'lowestPrice': '25905.1',
          'markedPrice': '25919.1',
          'openPrice': '26115.4',
          'prePositionFeeRate': '0.000011',
          'symbol': 'BTCUSDT',
          'turnover': '386249987.254913',
          'volume': '29237.4234'},
 'last': 25918.4,
 'low': 25905.1,
 'open': 26115.4,
 'percentage': -0.7543441800623386,
 'previousClose': None,
 'quoteVolume': 386249987.254913,
 'symbol': 'BTC/USDT:USDT',
 'timestamp': None,
 'vwap': 13210.808010356788}
```
```
p lbank2 fetchTicker "BTC/USDT"     
Python v3.10.9
CCXT v4.0.54
lbank2.fetchTicker(BTC/USDT)
{'ask': None,
 'askVolume': None,
 'average': None,
 'baseVolume': 6708.7767,
 'bid': None,
 'bidVolume': None,
 'change': None,
 'close': 25927.09,
 'datetime': '2023-08-22T16:00:49.670Z',
 'high': 26246.64,
 'info': {'symbol': 'btc_usdt',
          'ticker': {'change': '-0.1',
                     'high': '26246.64',
                     'latest': '25927.09',
                     'low': '25915.5',
                     'turnover': '174903571.67',
                     'vol': '6708.7767'},
          'timestamp': '1692720049670'},
 'last': 25927.09,
 'low': 25915.5,
 'open': None,
 'percentage': -0.1,
 'previousClose': None,
 'quoteVolume': 174903571.67,
 'symbol': 'BTC/USDT',
 'timestamp': 1692720049670,
 'vwap': 26070.859039025698}
```
```
p lbank2 fetchOrderBook "BTC/USDT" 2
Python v3.10.9
CCXT v4.0.54
lbank2.fetchOrderBook(BTC/USDT,2)
{'asks': [[25921.2, 0.9149], [25921.21, 1.2008]],
 'bids': [[25921.18, 1.3733], [25921.1, 2.0786]],
 'datetime': '2023-08-22T16:01:08.952Z',
 'nonce': None,
 'symbol': 'BTC/USDT',
 'timestamp': 1692720068952}
```
```
p lbank2 fetchOrderBook "BTC/USDT:USDT" 2
Python v3.10.9
CCXT v4.0.54
lbank2.fetchOrderBook(BTC/USDT:USDT,2)
{'asks': [[25924.7, 12.4309], [25924.8, 24.4475]],
 'bids': [[25924.6, 24.3214], [25924.5, 0.7493]],
 'datetime': '2023-08-22T16:01:20.387Z',
 'nonce': None,
 'symbol': 'BTC/USDT:USDT',
 'timestamp': 1692720080387}
```
```
p lbank2 fetchTime --swap                
Python v3.10.9
CCXT v4.0.54
lbank2.fetchTime()
1692720110653
```

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**